### PR TITLE
Validate Readme image paths

### PR DESCRIPTION
**Pull request by dansterenson**
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://github.com/demisto/etc/issues/33986
fixes: https://github.com/demisto/etc/issues/29002

## Description
1. Added a validation for the pack README file which the validate command did not check.
2. If a pack README file contains relative paths (which are not supported), the validate fails.
3. if a non pack README file contains relative paths which are not valid (doesn't exist), the validate fails.
4. if any README file contains absolute paths which are not valid (404 response), the validate fails.

## Must have
- [x] Tests
- [x] Documentation

### CODE REVIEW

File: demisto_sdk/commands/common/hook_validations/readme.py [hunk 1]
```diff
@@ -212,6 +218,95 @@ def is_image_path_valid(self) -> bool:
...
+        """ Validate readme images absolute paths - Check if absolute paths are not broken.
+
+        Arguments:
+            is_pack_readme (bool) - True if the the README file is a pack README, default: False
+
+        Returns:
+            list: List of the errors found
+        """
+        error_list = []
+        should_print_error = not is_pack_readme  # print error if its not pack readme
```
Comment (yaakovi) [inline demisto_sdk/commands/common/hook_validations/readme.py:286]: why? improve the comment

File: demisto_sdk/commands/common/hook_validations/readme.py [hunk 2]
```diff
@@ -212,6 +218,95 @@ def is_image_path_valid(self) -> bool:
...
+        absolute_links = re.findall(
+            r'(\!\[.*?\]|src\=)(\(|\")(https://(raw\.githubusercontent|github).com/demisto/content/.*?)(\)|\")', self.readme_content,
```
Comment (yaakovi) [inline demisto_sdk/commands/common/hook_validations/readme.py:288]: what if the image is taken from somewhere other than Github? or other than content repo (premium/gold)

File: demisto_sdk/commands/common/hook_validations/readme.py [hunk 3]
```diff
@@ -212,6 +218,95 @@ def is_image_path_valid(self) -> bool:
...
+            re.IGNORECASE)
+
+        if absolute_links:
+            for link in absolute_links:
+                try:
+                    img_url = link[2]
+                except IndexError:
+                    continue
+                img_url = img_url.strip('()')
```
Comment (yaakovi) [inline demisto_sdk/commands/common/hook_validations/readme.py:297]: according to the regex, there shouldn't be anything to strip

File: demisto_sdk/commands/common/hook_validations/readme.py [hunk 4]
```diff
@@ -212,6 +218,95 @@ def is_image_path_valid(self) -> bool:
...
+        should_print_error = not is_pack_readme  # print error if its not pack readme
+        relative_images = re.findall(r'(\!\[.*?\])(\((?!http).*?\))', self.readme_content, re.IGNORECASE)
+        if relative_images:
+            if is_pack_readme:  # Check whether the README type is pack README which do not support relative paths.
+                for img in relative_images:
+                    try:
+                        relative_path = img[1]
+                    except IndexError:
+                        continue
+                    if relative_path:
```
Comment (yaakovi) [inline demisto_sdk/commands/common/hook_validations/readme.py:254]: `text`
will not fail the validation

File: demisto_sdk/commands/common/hook_validations/readme.py [hunk 5]
```diff
@@ -212,6 +218,95 @@ def is_image_path_valid(self) -> bool:
...
+                        error_list.append(formatted_error)
+
+            else:  # Not pack readme relative paths are allowed but have to be valid.
+                for img in relative_images:
+                    try:
+                        relative_path = img[1]
+                    except IndexError:
+                        continue
+                    # playbook doc generator automatically adds 'Insert the link to your image here' to the
+                    # readme image path, which we want to skip.
```
Comment (yaakovi) [inline demisto_sdk/commands/common/hook_validations/readme.py:267]: why skip? it should also fail the validation

File: demisto_sdk/commands/common/hook_validations/readme.py [hunk 6]
```diff
@@ -212,6 +218,95 @@ def is_image_path_valid(self) -> bool:
...
+                try:
+                    response = requests.get(img_url, verify=False, timeout=10)
+                    if response.status_code != 200:
+                        error_message, error_code = Errors.invalid_readme_image_absolute_path_error(img_url)
+                        formatted_error = \
+                            self.handle_error(error_message, error_code, file_path=self.file_path,
+                                              should_print=should_print_error)
+                        error_list.append(formatted_error)
+                except requests.HTTPError as ex:
+                    raise Exception(f'Failed reaching to integration doc link {img_url} - {ex}')
```
Comment (yaakovi) [inline demisto_sdk/commands/common/hook_validations/readme.py:307]: who will catch this error?

File: demisto_sdk/commands/common/errors.py [hunk 7]
```diff
@@ -1342,6 +1345,24 @@ def image_path_error(path, alternative_path):
...
+    def pack_readme_image_relative_path_error(path):
+        return f'Detected following image: {path} which is ' \
+               f'not using an absolute url (url does not start with http/https). See ' \
+               f'https://xsoar.pan.dev/docs/integrations/integration-docs#images for further info on how to ' \
+               f'add images to README files.'
+
+    @staticmethod
+    @error_code_decorator
+    def invalid_readme_image_relative_path_error(path):
+        return f'The following relative path is not valid - {path} please recheck it.'
```
[the remark this answers is not in the document]
Comment (dansterenson) [reply demisto_sdk/commands/common/errors.py:1359]: It handles each error separately, each link gets the its own error.

[the remark this answers is not in the document]
Comment (dansterenson) [reply demisto_sdk/commands/common/errors.py:1359]: image

File: demisto_sdk/commands/common/errors.py [hunk 8]
```diff
@@ -1342,6 +1345,24 @@ def image_path_error(path, alternative_path):
...
+
+    @staticmethod
+    @error_code_decorator
+    def invalid_readme_image_absolute_path_error(path):
+        return f'The following link seems to be broken, please repair it - {path}'
```
[the remark this answers is not in the document]
Comment (dansterenson) [reply demisto_sdk/commands/common/errors.py]: Done

File: demisto_sdk/commands/common/hook_validations/readme.py (hunk 1 above)
Comment (dansterenson) [reply demisto_sdk/commands/common/hook_validations/readme.py]: Done

File: demisto_sdk/commands/common/hook_validations/readme.py (hunk 2 above)
Comment (dansterenson) [reply demisto_sdk/commands/common/hook_validations/readme.py]: Done

File: demisto_sdk/commands/common/hook_validations/readme.py (hunk 3 above)
Comment (dansterenson) [reply demisto_sdk/commands/common/hook_validations/readme.py]: Done

File: demisto_sdk/commands/common/hook_validations/readme.py (hunk 4 above)
Comment (dansterenson) [reply demisto_sdk/commands/common/hook_validations/readme.py]: Fixed

File: demisto_sdk/commands/common/hook_validations/readme.py (hunk 5 above)
Comment (dansterenson) [reply demisto_sdk/commands/common/hook_validations/readme.py]: Done

File: demisto_sdk/commands/common/hook_validations/readme.py (hunk 6 above)
Comment (dansterenson) [reply demisto_sdk/commands/common/hook_validations/readme.py]: Done

File: demisto_sdk/commands/common/hook_validations/readme.py [hunk 9]
```diff
@@ -212,6 +218,108 @@ def is_image_path_valid(self) -> bool:
...
+        Arguments:
+            is_pack_readme (bool) - True if the the README file is a pack README, default: False
+
+        Returns:
+            list: List of the errors found
+        """
+        error_list = []
+        error_code: str = ''
+        error_message: str = ''
+        should_print_error = not is_pack_readme  # If error was found, print it only if its not pack readme.
```
Comment (yaakovi) [inline demisto_sdk/commands/common/hook_validations/readme.py:247]: code is self-explanatory,
comment should say ***why*** we do it

File: demisto_sdk/commands/common/hook_validations/readme.py [hunk 10]
```diff
@@ -212,6 +218,108 @@ def is_image_path_valid(self) -> bool:
...
+        relative_images = re.findall(r'(\!\[.*?\])\(((?!http).*?)\)$', self.readme_content, re.IGNORECASE | re.MULTILINE)
+        relative_images += re.findall(  # HTML image tag
+            r'(src\s*=\s*\"((?!http).*?)\")', self.readme_content,
+            re.IGNORECASE | re.MULTILINE)
+
+        if relative_images:
```
Comment (yaakovi) [inline demisto_sdk/commands/common/hook_validations/readme.py:253]: condition is redundant.
if the list is empty, the loop will not do anything (no iterations).
if the list is not empty, the loop will work as regularly.

File: demisto_sdk/commands/common/hook_validations/readme.py [hunk 11]
```diff
@@ -212,6 +218,108 @@ def is_image_path_valid(self) -> bool:
...
+        Returns:
+            list: List of the errors found
+        """
+        error_list = []
+        error_code: str = ''
+        error_message: str = ''
+        should_print_error = not is_pack_readme  # If error was found, print it only if its not pack readme.
+        relative_images = re.findall(r'(\!\[.*?\])\(((?!http).*?)\)$', self.readme_content, re.IGNORECASE | re.MULTILINE)
+        relative_images += re.findall(  # HTML image tag
+            r'(src\s*=\s*\"((?!http).*?)\")', self.readme_content,
```
Comment (yaakovi) [inline demisto_sdk/commands/common/hook_validations/readme.py:250]:
```suggestion
            r'(src\s*=\s*"((?!http).*?)")', self.readme_content,
```

File: demisto_sdk/commands/common/hook_validations/readme.py [hunk 12]
```diff
@@ -212,6 +218,108 @@ def is_image_path_valid(self) -> bool:
...
+            re.IGNORECASE | re.MULTILINE)
+
+        if relative_images:
+            for img in relative_images:
+                try:
+                    # striping in case there are whitespaces at the beginning/ending of url.
+                    prefix = '' if 'src' in img[0] else img[0].strip()
+                    relative_path = img[1].strip()
+                except IndexError:
+                    continue
```
Comment (yaakovi) [inline demisto_sdk/commands/common/hook_validations/readme.py:260]: I'm not sure what's going on here, can you explain?

File: demisto_sdk/commands/common/hook_validations/readme.py [hunk 13]
```diff
@@ -212,6 +218,108 @@ def is_image_path_valid(self) -> bool:
...
+                if 'insert URL to your image' in relative_path:
+                    # some old README files with HTML img tags contain this line but there are in a comment and
+                    # can be skipped
+                    continue
```
Comment (yaakovi) [inline demisto_sdk/commands/common/hook_validations/readme.py:264]: how many of these kinds of files do we have?
better to fix them as well

File: demisto_sdk/commands/common/hook_validations/readme.py [hunk 14]
```diff
@@ -212,6 +218,108 @@ def is_image_path_valid(self) -> bool:
...
+                    # the line is generated automatically in playbooks readme, the user should replace it with
+                    # an image or remove the line.
+                    error_message, error_code = Errors.invalid_readme_image_error(prefix + f'({relative_path})',
+                                                                                  error_type='insert_image_link_error')
+                elif is_pack_readme:
+                    error_message, error_code = Errors.invalid_readme_image_error(prefix + f'({relative_path})',
+                                                                                  error_type='pack_readme_relative_error')
+                else:
+                    # generated absolute path from relative and checked for the file existence.
+                    if not os.path.exists(os.path.join(self.file_path.parent, relative_path)):
```
Comment (yaakovi) [inline demisto_sdk/commands/common/hook_validations/readme.py:275]:
```suggestion
                    # generates absolute path from relative and checks for the file existence.
                    if not os.path.isfile(os.path.join(self.file_path.parent, relative_path)):
```

File: demisto_sdk/commands/common/hook_validations/readme.py [hunk 15]
```diff
@@ -212,6 +218,108 @@ def is_image_path_valid(self) -> bool:
...
+            list: List of the errors found
+        """
+        error_list = []
+        should_print_error = not is_pack_readme  # pack readme errors are handled and printed during the pack unique
+        # files validation.
+        absolute_links = re.findall(
+            r'(!\[.*\])\((https://.*)\)$', self.readme_content, re.IGNORECASE | re.MULTILINE)
+        absolute_links += re.findall(  # image tag
+            r'(src\s*=\s*"(https://.*?)")', self.readme_content, re.IGNORECASE | re.MULTILINE)
+        if absolute_links:
```
Comment (yaakovi) [inline demisto_sdk/commands/common/hook_validations/readme.py:301]: same here

File: demisto_sdk/commands/common/hook_validations/readme.py (hunk 9 above)
Comment (dansterenson) [reply demisto_sdk/commands/common/hook_validations/readme.py:247]: Done

File: demisto_sdk/commands/common/hook_validations/readme.py (hunk 10 above)
Comment (dansterenson) [reply demisto_sdk/commands/common/hook_validations/readme.py:253]: fixed

File: demisto_sdk/commands/common/hook_validations/readme.py (hunk 11 above)
Comment (dansterenson) [reply demisto_sdk/commands/common/hook_validations/readme.py:250]: fixed

File: demisto_sdk/commands/common/hook_validations/readme.py (hunk 14 above)
Comment (dansterenson) [reply demisto_sdk/commands/common/hook_validations/readme.py:275]: fixed

File: demisto_sdk/commands/common/hook_validations/readme.py (hunk 15 above)
Comment (dansterenson) [reply demisto_sdk/commands/common/hook_validations/readme.py:301]: fixed

File: demisto_sdk/commands/common/hook_validations/readme.py (hunk 12 above)
Comment (dansterenson) [reply demisto_sdk/commands/common/hook_validations/readme.py:260]: Fixed

File: demisto_sdk/commands/common/hook_validations/readme.py (hunk 13 above)
Comment (dansterenson) [reply demisto_sdk/commands/common/hook_validations/readme.py:264]: fixed those files and removed the if statement.

File: demisto_sdk/commands/common/errors.py [hunk 16]
```diff
@@ -1364,22 +1362,32 @@ def image_path_error(path, alternative_path):
...
+
+    @staticmethod
+    @error_code_decorator
+    def invalid_readme_image_error(path, error_type):
+        return 'Error in readme image:\n' + {
+            'pack_readme_relative_error': Errors.pack_readme_image_relative_path_error,
+            'general_readme_relative_error': Errors.invalid_readme_image_relative_path_error,
+            'general_readme_absolute_error': Errors.invalid_readme_image_absolute_path_error,
+            'insert_image_link_error': Errors.invalid_readme_insert_image_link_error
+        }[error_type](path)
```
[the remark this answers is not in the document]
Comment (dansterenson) [reply demisto_sdk/commands/common/errors.py:1390]: Done

**Review by yaakovi — APPROVED** [review]

**Review by reutshal — APPROVED** [review]

**Pull request merged**